### PR TITLE
feat: keybinds by oen | (limited) Part 1

### PR DESCRIPTION
**Pull request by InnerCircleTFS**
# Credits : oen
I’m not interested in the fan war between OTC, I liked this feature and I want it in my client. I’m sharing it.

Here is the original code for keybinds  of Oen.

- Part 1 integrates Keybins **WITHOUT the walk**. I could adapt it, but my **intention is not to delay the Android branch any further**, as it would cause conflicts in that branch. (game_joystick, game_walk)
I will adapt Part 2 when that branch is merged in main repo.

- **Hotkeys are not adapted** because I like the old version(me server is 10.98). If someone wants to adapt it in the future, I would appreciate it if they kept the old version compatible. :)

- **the "save" functionality is not adapted** either, because I didn't like it. (if someone wants to do it, he should add a button and call this `function applyChangedOptions()` and `setOption(key, option.value, option.force))`

# Keybinds system

## Notes

- Arrow keys are reserved and can NOT be used
- Default keybinds can NOT share keys (you will be warned if it occurs)
- Not every keybind is currently added, contributions in that matter are welcome
- Some OTC keybinds are bound to UI/UX and should not be part of this Keybinds system
- Action Bar is currently (yet) not integrated

## Creating new keybind

Best place to add is init (or whatever onLoad executes) of your module.

```lua
Keybind.new(category, action, primary, secondary, alone)

```

## Params

- category (string) - category name, eg. "Dialogs"
- action (string) - action name, eg. "Open Options" (creating "Dialogs: Open Options" name)
- primary (string | table) - default primary key to trigger action, if used as string (can be empty but not nil), it will be applied to both Chat Modes, otherwise use table:

```
{
  [CHAT_MODE.ON] = "key", (can be empty)
  [CHAT_MODE.OFF] = "key" (can be empty)
}
```

- secondary (string | table) - same as primary key
- alone (bool, optional) - if true, action can be bound to only single key without modifiers (Ctrl, Alt, Shift)

```lua
Keybind.new("Dialogs", "Open Options", "Ctrl+O", "")

Keybind.new("Chat", "Send current chat line", { [CHAT_MODE.ON] = "Enter", [CHAT_MODE.OFF] = "" }, "") -- table for primary, string for secondary

Keybind.new("Movement", "Go North", {[CHAT_MODE.ON] = "", [CHAT_MODE.OFF] = "W"}, {[CHAT_MODE.ON] = "Numpad8", [CHAT_MODE.OFF] = ""}, true)
```

## Binding keybinds

Creating keybinds is not enough, you have to make them do something. Binding is separated for multiple reasons but the most important one is that not everything is ready and loaded when client starts, some modules require player to enter the game to bind the keys. Must be used AFTER `Keybind.new.`

`Keybind.bind(category, action, callbacks, widget)`

# Params

- category (string) - category name, eg. "Dialogs"
- action (string) - action name, eg. "Open Options"
- callbacks (table) - table with table of config values:

```lua
{
  {
    type = (KEY_DOWN | KEY_UP | KEY_PRESS),
    callback = (function),
    alone = (bool)
  },
  -- can be multiple (see game_walking)
  {
    type = (KEY_DOWN | KEY_UP | KEY_PRESS),
    callback = (function),
    alone = (bool)
  }
}
```

- widget (UIWidget, optional) - widget that should listen to key events (if not provided it will default to Root)

## Example

```lua
  Keybind.bind("UI", "Toggle Full Map", {
    {
      type = KEY_DOWN,
      callback = toggleFullMap,
    }
  })

  Keybind.bind("Movement", "Go " .. DirectionString[North], {
    {
      type = KEY_DOWN,
      callback = function() changeWalkDir(dir) end,
      alone = true
    },
    {
      type = KEY_UP,
      callback = function() changeWalkDir(dir, true) end,
      alone = true
    },
    {
      type = KEY_PRESS,
      callback = function(c, k, ticks) smartWalk(dir, ticks) end
    }
  }, gameRootPanel)
```

## Deleting Keybinds

Best place to add is terminate (or whatever onUnload executes) of your module. This is required, altho there is a check to prevent duplicate Keybinds, it is better to delete unused one before they readded in init or somewhere.

```lua
Keybind.delete(category, action)
```

## Params

- category (string) - category name, eg. "Dialogs"
- action (string) - action name, eg. "Open Options"

## Example

```lua
Keybind.delete("Dialogs", "Open Options")
```

# Credits : oen
I’m not interested in the fan war between OTC, I liked this feature and I want it in my client. I’m sharing it.

### CODE REVIEW

**Comment by InnerCircleTFS** [comment]
Notes:
There were minimal errors that pr fixed:

- Shader and RuleViolation had the same hotkeys (Ctrl + Y), so I took the liberty of changing RuleViolation to Ctrl + U.
- While testing, I noticed error in module game_cooldown that caused the console to lose focus.
- When creating a button with

```css
align-right: true
spacing: -5
```
The negative sign wasn't considered because it was uint8_t, so I changed it to int8_t.


If someone else could test it, I would appreciate it.

**Review by brkihel — APPROVED** [review]

File: modules/client_options/options.lua [hunk 1]
```diff
@@ -153,6 +149,12 @@ local function setupComboBox()
         end
     end
 
+    for _, preset in ipairs(Keybind.presets) do
+        listKeybindsPanel:addOption(preset)
+    end
+    listKeybindsPanel.onOptionChange = function(comboBox, option)
+        setOption('listKeybindsPanel', option)
+    end
```
Comment (kokekanon) [inline modules/client_options/options.lua:157]: I spent a good while thinking about why it wasn't working; when opening the client, it doesn't load the last saved keybinds.

```suggestion
    for _, preset in ipairs(Keybind.presets) do
        listKeybindsPanel:addOption(preset)
    end
    listKeybindsPanel.onOptionChange = function(comboBox, option)
        setOption('listKeybindsPanel', option)
    end
    panels.keybindsPanel.presets.list:setCurrentOption(Keybind.currentPreset)
```

File: modules/client_options/keybins.lua [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+    local chatMode = getChatMode()
+    local preset = panels.keybindsPanel.presets.list:getCurrentOption().text
+    if Keybind.hotkeys[chatMode][preset] then
+        for _, hotkey in ipairs(Keybind.hotkeys[chatMode][preset]) do
+            addHotkey(hotkey.hotkeyId, hotkey.action, hotkey.data, hotkey.primary, hotkey.secondary)
+        end
+    end
+end
+
+
```
Comment (kokekanon) [inline modules/client_options/keybins.lua]:
```suggestion
```

File: modules/client_options/keybins.lua [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+    local widget = panels.keybindsPanel:recursiveGetChildById('list')
+    presetOption(widget, 'currentPreset', value, false)
+    changedKeybinds = {}
+    changedHotkeys = {}
+    applyChangedOptions()
+    updateKeybinds()
+end
+
+function debug()
+    local currentOptionText = panels.keybindsPanel.presets.list:getCurrentOption().text
```
Comment (kokekanon) [inline modules/client_options/keybins.lua]:
```suggestion
    local currentOptionText = Keybind.currentPreset
```

File: modules/client_options/keybins.lua [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+        end
+        keyEditWindow:destroy()
+        keyEditWindow = nil
+      end
+    
+      actionSearchEvent = nil
+end
+
+function listKeybindsComboBox(value)
+    local widget = panels.keybindsPanel:recursiveGetChildById('list')
```
Comment (kokekanon) [inline modules/client_options/keybins.lua]:
```suggestion
    local widget = panels.keybindsPanel.presets.list
```

File: modules/client_options/keybins.lua [hunk 5]
```diff
@@ -0,0 +1,766 @@
+
```
Comment (kokekanon) [inline modules/client_options/keybins.lua]:
```suggestion
```

**Review by kokekanon — COMMENTED** [review]
parcial review

**Comment by kokekanon** [comment]
that is visible in 8.6 change from onInit() to onGameStart()
image

**Comment by kokekanon** [comment]
photoshop
quest log no work
and fps/ ping no work
image

**Comment by mehah** [comment]
@kokekanon 
@InnerCircleTFS done?

**Comment by InnerCircleTFS** [comment]
> @kokekanon @InnerCircleTFS done?

`cee90ec` fix bugs found by kokekanon

 works in my environment

**Review by kokekanon — APPROVED** [review]
works so far, 

only tested on windows. 

no test in web client

**Pull request merged**